### ISSUE TITLE
Write up a "how to contribute what, where", to clear up misunderstandings about what code would be accepted.

### ISSUE THREAD

**Issue by Pomax** [processing-js/processing-js#148]
There is clearly a lot of misunderstanding about what is and is not accepted in processing.js as a project, and a document that clearly states how the project falls apart (e.g. core vs. extensions) and "how to contribute what kind of code, where" will go a long way towards clearing those up.

Recent bugs that touch this subject are #140 and #145, starting with the text in those and creating a new document on the repo (and possibly the site) will be an important first step.

**Comment by teo1978** [comment]
In reply to your last comment in https://github.com/processing-js/processing-js/issues/145

> The idea so far has been if you write a dual technology library, then you write your code against the java version as you'd do already, and then simply include the javascript version as an on-page <script> for your page, and things should just work.

Am I missing something, or am I right in thinking that that would make the workflow a bit painful?
When you are writing a sketch in Processing in JavaScript mode, you usually continuously edit it in the PDE, and every time you need to test the last three lines you've typed you hit the play button. That generates the index.html in the web-export folder. (actually, that's regenerated every time you save)

If you have to add a library by editing the html (ignoring the fact that this is not as user-friendly for newcomers as the install-once-import-whenever-you-want way of using libraries), then you have to forget about using the play button from the PDE. You'll have to copy the web-export folder to somewhere else, and every time you save changes to the sketch, copy the .pde to that folder. Or edit the .pde directly in that other folder (ignoring the prompt about creating a folder with the same name for the sketch, and/or ending up with a garbage additional web-export subfolder), but additionally you'll need a local web server other than the built-in one that the Processing IDE's JavaScript mode provides. And forget about the Play button.

Or viceversa, make a backup copy of your edited html. Every time you save the sketch, restore the backup of the edited html into the web-export folder. This is if you want to test the sketch easily with the play button from the IDE.

**Comment by teo1978** [comment]
In reply to your last comment in https://github.com/processing-js/processing-js/issues/145

> The idea so far has been if you write a dual technology library, then you write your code against the java version as you'd do already, and then simply include the javascript version as an on-page <script> for your page, and things should just work.

Am I missing something, or am I right in thinking that that would make the workflow a bit painful?
When you are writing a sketch in Processing in JavaScript mode, you usually continuously edit it in the PDE, and every time you need to test the last three lines you've typed you hit the play button. That generates the index.html in the web-export folder. (actually, that's regenerated every time you save)

If you have to add a library by editing the html (ignoring the fact that this is not as user-friendly for newcomers as the install-once-import-whenever-you-want way of using libraries), then you have to forget about using the play button from the PDE. You'll have to copy the web-export folder to somewhere else, and every time you save changes to the sketch, copy the .pde to that folder. Or edit the .pde directly in that other folder (ignoring the prompt about creating a folder with the same name for the sketch, and/or ending up with a garbage additional web-export subfolder), but additionally you'll need a local web server other than the built-in one that the Processing IDE's JavaScript mode provides. And forget about the Play button.

Or viceversa, make a backup copy of your edited html. Every time you save the sketch, restore the backup of the edited html into the web-export folder. This is if you want to test the sketch easily with the play button from the IDE.

**Comment by teo1978** [comment]
I don't understand why only a small piece of my comment shows up. If I edit it I can see it in its entirety

**Comment by GoToLoop** [comment]
@Pomax seems to ignore the Processing's target audience is "artists"!
What we want is cross-mode sketches that work both in Java & JS modes!

**Comment by Pomax** [comment]
one thing to remember is that the Processing and Processing.js projects are basically completely unaffiliated, and how Processing.js does things is completely independent of what the Processing devs do with their "javascript mode" conversion.

The description of how to load additional libraries is based on how Processing.js and JavaScript work in the browser, so if the PDE is not generating the appropriate html code for getting additional JS libraries loaded in, then that's worth filing a bug for on the processing repository along the lines of "when we work in javascript mode, let us pick additional .js files that need to be loaded", which would solve that problem.

**Comment by Pomax** [comment]
@GoToLoop can you please tone it down? You're being very accusatory against me, personally, accusing me of not doing things in projects I don't even work on. That hurts, and doesn't help me or you get things done to improve Processing.js. It doesn't help to accuse me of lots of things that are not even in my realm of control.

If you want to help improve the interoperation between two *completely independent* projects, that's fantastic, but you're demanding seemless integration between Processing's PDE and Processing.js, even though the "javascript mode" that the PDE offers has *nothing to do* with Processing.js, other than "that's what the Processing team decides to use themselves". It's not a collaboration between Processing and Processing.js, it's just a decision the Processing team made, to use Processing.js for an export-for-the-web functionality, that isn't very future proof because assumptions were made on how things work, and now that's simply what you get in the PDE. 

I don't work on Processing, I don't work on the PDE, and other than the occasional tweet at each other about developments in the two projects, there is zero collaboration between the two projects.

So: I understand that you want seemless integration. Who doesn't? But it's unreasonable to expect me to be able to do that for you: I don't control what happens in the Processing project in the slightest, I am not a contributor on that project, and I can only tell you how Processing.js does things.

Don't let the names of the projects fool you: We're still talking about two *independent* projects, so if you have good ideas for Processing.js, let's work on those, but if your ideas involve "changing what the Processing PDE does", then really: I am not the guy to talk to, I have no power over it, I have no knowledge of its codebase, I have nothing to contribute there. You want to file those issues over on [the processing repository](https://github.com/processing/processing)

**Comment by GoToLoop** [comment]
So sorry! I thought you were the developer or what was left of this project!  =(

**Comment by teo1978** [comment]
The misunderstanding is that "JavaScript mode for Processing" (as a mechanism to export a Processing sketch using Processing.js) is not part of Processing.js itself. I was fooled too (not until now but until little ago).

**Comment by GoToLoop** [comment]
> One thing to remember is that the Processing and Processing.js projects are basically completely unaffiliated, and how Processing.js does things is completely independent of what the Processing devs do with their "javascript mode" conversion.

That's exactly why I'm advocating for the "abandoned" pjs framework to rise above its 100 subservience and have more extras of its own! It's a pity to let such cool project to die out the way it is!

Sorry again for "accusing" you. You pretty seemed the 1 in charge here!  @_@

**Comment by GoToLoop** [comment]
> The misunderstanding is that "JavaScript mode for Processing" (as a mechanism to export a Processing sketch using Processing.js) is not part of Processing.js itself.

I'm aware of that too! But w/o JS Mode I wouldn't even know that pjs framework existed at all!
However, modifying API like placing a callback in **requestImage()** is framework related!

**Comment by teo1978** [comment]
@GoToLoop the extras you're advocating for belong to JavaScript mode, not Processing.js actually.

Who are letting such a cool project die (and this includes both PJS and P5 itself) are the Processing developer team, starting from the moment they dropped support for exporting java applets (understandably, given how the java plugin has become an obsolete and shitty platform for web) and thought that building a half-made bridge with Processing JS via the JavaScript mode and without any effort in maintaining it (nor in building any collaboration with PJS developer[s]), would be enough to keep publishing Processing sketches on the web a viable thing. Not to mention their lack of interest in designing asynchronous APIs and improve the API in general (getting rid of conventionally-named methods in the sketch as callbacks in favour of passing callback functions as parameters, something perhaps plan to do in 2035)

**Comment by GoToLoop** [comment]
Are you sure that modifying an API is JS Mode related? JS Mode doesn't modify the framework but only the generated HTML!  O_o

**Comment by teo1978** [comment]
Modifying the API isn't, but that's the part that a library could already do without modifying Processing.js

**Comment by teo1978** [comment]
Anyway, now that I've discovered P5*js, I just hope it takes off so that we all can forget about Processing. Processing was absolutely glorious, but the sad truth is it's dead. I hoped it would resurrect thanks to PJS but this definitely clarifies that's not gonna happen:
> other than the occasional tweet at each other about developments in the two projects, there is zero collaboration between the two projects.

**Comment by Pomax** [comment]
I'm looking forward to P5.js become the new target for Processing. Especially with Shiffman's involvement, I suspect it may go far if they keep at it.

**Comment by GoToLoop** [comment]
p5*js rejuvenates the whole thing. However many Processing folks can't grasp JS' syntax for OOP just now!
Next JS version 6 (Harmony) is gonna bring classes to the language and fixes for the `this` keyword.
It's gonna take a awhile before all main browsers adopt the next version though!

**Comment by teo1978** [comment]
I know we're drifting OT again here, but couldn't p5*js just provide some class-like framework, i.e. functions that emulate "creating classes" and "adding methods to them", for example with techniques similar to those described in the book "JavaScript: the good parts" or something like that? Instead of waiting until JS brings actual classes to the language.

**Comment by GoToLoop** [comment]
* p5*js is JS syntax the same way p5 is for Java syntax.
* It won't force any artificial JS constructs besides what's offered by JS language.
* Although we might try to mix p5*js w/ another framework which does that!
* Eiher we w8 for when Java syntax arrives for JS or use some sugar syntax language like CoffeeScript.

**Comment by teo1978** [comment]
> It won't force any artificial JS constructs besides what's offered by JS language.
> Eiher we w8 for when Java syntax arrives for JS or use some sugar syntax language like CoffeeScript or TypeScript.

I wasn't talking about any artificial constructs besides JS, nor any kind of pre-compilation. Just pure framwork, done in pure JavaScript. Just pure framework, written un pure JS. That's possible.

Obviously you'll never be able to write
```
class MyClass extends MyParentClass {
  function myMethod() { ...}
}
```
but it's perfectly possible (and I'm pretty sure someone has done it already, just too lazy to search) to create a framework that will allow you to write something like

```
var MyClass=SomeClassFactory.createClass(MyParentClass, {
  myMethod: function() {...}
});
```
or probably something nicer.

**Comment by GoToLoop** [comment]
I find something like `var MyClass=SomeClassFactory.createClass(MyParentClass, { // ... });`
too much verbose that JS' `prototype` way is more preferable!
For Java-like class syntax, go w/ CoffeeScript: http://CoffeeScript.org.  ;-)

**Comment by teo1978** [comment]
@Pomax I've just come across an use case where I needed to override a core function (in my case loadXML), and I remembered you saying how easy it would be to extend the framework without touching it, by just adding extra javascript files, and guess what? It's next to (if not completely) impossible.

Sure, adding a library function that can be called from the sketch is easy. As long as you don't need it to work as a "method" of the sketch "class". And as long as you don't need to override an existing function of Processing.

That *could* be made easier if the way Processing.js instantiates a sketch was modified.

Let's consider an example.
I wanted to write a javascript file, call it loadxml-patch.js, which would work as a hack to fix https://github.com/processing-js/processing-js/issues/150 (on existing versions of Processing.js)

It turns out that in JavaScript mode, any .js file included in the sketch folder is automatically copied to the web-export folder and added as a \<script\> tag in index.html after the Processing.js one.

So, if only I could write a .js file that overrides the behavior of loadXML(), I could provide to my students a  "patch" (no matter how hacky) that fixes #150. It would only require them to copy this .js file into their sketch's folder.

However, there's no javascript code (or if there is it is extremely convoluted) that can be written to replace loadXML (and the same applies to nearly any core function) with a different implementation.

I can't write something like
```
Processing.something.prototype.loadXML = function (uri) { ...}
```
Nor does even this work:
```
Processing.getInstanceById(getProcessingSketchId()).loadXML=function(){...}
```
and I wonder if any way exists at all.

I'm pretty sure the code that defines the sketch methods (and the code that initializes a sketch putting those methods into it) could be rewritten in such a way that one could easily override core functions, or add new ones (whenever they need to behave like "methods" attached to the sketch and can't simply be written as global functions), with some code similar to my first example above.

**Comment by Pomax** [comment]
yeah I remember we had function hooks -somewhere- that made that a lot easier, but I don't off the top of my head remember whether those ever landed or whether they were part of work that was done to make "plugins" (written in JS) easier. I'll see if I can find that code this weekend, it'll be worth putting that in if it never landed. And if I can't find it, it'll be worth writing that back up to make extensibility easier.

**Comment by fjenett** [comment]
Re: https://github.com/processing-js/processing-js/issues/148#issuecomment-77709549

This can simply be done using a custom template folder (link to create one is in the menu) with your sketch. Edit the template.html in there and it will be used instead of the default template file.

Please create an issue over at my repo if you have trouble with that:
https://github.com/fjenett/javascript-mode-processing

**Comment by Pomax** [comment]
hmm, Interesting. I don't see any PFont errors when running the font/text ref tests, but that might be because it's not using the code you are - do you have a snippet of code I can add to the ref tests for verifying and hunting this down?

**Comment by Pomax** [comment]
nice. In that case I will close this issue, with a [new issue filed](https://github.com/processing-js/processing-js/issues/165) for surfacing and documenting the Processing hooks for libraries that need to reach into Pjs itself instead of offering additional JS library functions and objects.
